### PR TITLE
feat: Frontend work to support multiple generalized applications

### PR DESCRIPTION
**Pull request by kristenarmes**
Signed-off-by: Kristen Armes <karmes@lyft.com>

### Summary of Changes

This frontend change supports the changes made in https://github.com/amundsen-io/amundsen/pull/1577. Currently, there is a link that will be displayed if there is a table writer application in the table metadata but the page does not support multiple applications. This change modifies the link to become a dropdown instead. It will take the existing `table_writer` field and the new optional `table_apps` field, and it will use the information to render one or more dropdown components where the link used to be depending on the amount of types of applications included in the list.

This will support Airflow and Databricks as their own dropdowns since those were the previously supported applications. Any that fall outside of those two would be included in a separate generic dropdown list.

The dropdown(s) will group the applications by their `kind`, such as Producing or Consuming, and then by `id` within those groupings. Any applications in the generic dropdown will be grouped by `name`, then `kind`, then `id`.

### Tests

Added JS tests for the `ApplicationDropdown`.

### Documentation

N/A

### CheckList
Make sure you have checked **all** steps below to ensure a timely review.
- [X] PR title addresses the issue accurately and concisely. Example: "Updates the version of Flask to v1.0.2"
    - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
- [X] PR includes a summary of changes.
- [ ] PR adds unit tests, updates existing unit tests, __OR__ documents why no test additions or modifications are needed.
- [X] In case of new functionality, my PR adds documentation that describes how to use it.
    - All the public functions and the classes in the PR contain docstrings that explain what it does

### CODE REVIEW

File: frontend/amundsen_application/proxy/issue_tracker_clients/jira_client.py [hunk 1]
```diff
@@ -160,7 +162,7 @@ def create_issue(self,
             }, labels=self.issue_labels,
                 summary=title,
                 description=(f'{description} '
-                             f'\n *Reported By:* {user_email} '
+                             f'\n *Reported By:* {reporting_user_str if reporting_user_str else user_email} '
```
Comment (kristenarmes) [inline frontend/amundsen_application/proxy/issue_tracker_clients/jira_client.py:165]: This is a small unrelated change to follow up on previous frontend Jira work. This will link the profile url for the user who created the Jira ticket rather than only displaying their email.

File: frontend/amundsen_application/static/js/interfaces/TableMetadata.ts [hunk 2]
```diff
@@ -102,6 +103,7 @@ export interface TableMetadata {
   name: string;
   description: string;
   table_writer: TableWriter;
```
Comment (dkunitsk) [inline frontend/amundsen_application/static/js/interfaces/TableMetadata.ts:105]: would it be difficult to change the name from TableWriter to App/TableApp for the benefit of future maintainers.

File: frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/constants.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
+export const APPLICATIONS_LABEL = 'Applications';
```
Comment (dkunitsk) [inline frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/constants.ts:1]: nit about the filename: WriterDropdown to AppDropdown or some such

File: frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/index.tsx [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+  return (
+    <MenuItem
+      className={menuItemClassName}
+      href={app.application_url}
+      onClick={logClick}
+      target="_blank"
+    >
+      <div className="writer-dropdown-menu-item-row">
+        <span className="menu-item-content">
+          {app.id ? app.id : NONE_VALUE}
```
Comment (dkunitsk) [inline frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/index.tsx:83]: there will always be an app id

File: frontend/amundsen_application/static/images/icons/Application.svg [hunk 5]
```diff
@@ -0,0 +1,4 @@
+<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (Golodhros) [inline frontend/amundsen_application/static/images/icons/Application.svg:1]: What does this icon look like?

I would expect it to be named 'application.svg' instead.

File: frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/index.tsx [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+const getMenuItem = (app: TableWriter, menuItemClassName) => {
+  if (app.name.toLowerCase() === AIRFLOW) {
+    const [dagId, ...task] = app.id.split('/');
+    const taskId = task.join('/');
+    return (
+      <MenuItem
+        className={menuItemClassName}
+        href={app.application_url}
+        onClick={logClick}
+        target="_blank"
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/index.tsx:61]: Any link that has target="_blank" needs to have rel="noopener noreferrer" for security reasons.
I see this is not exactly a link, but I guess is kind of?

File: frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/index.tsx [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+      </MenuItem>
+    );
+  }
+  return (
+    <MenuItem
+      className={menuItemClassName}
+      href={app.application_url}
+      onClick={logClick}
+      target="_blank"
+    >
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/index.tsx:80]: Nit: It seems this gets reused on both cases, so I would keep it here and do something like:
```
<MenuItem
      className={menuItemClassName}
      href={app.application_url}
      onClick={logClick}
      target="_blank"
    >
        {isAirflow ? (
           //airflow menu-item-rows here
        ) : (
           //non-airflow menu item row
        )}
```

File: frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/index.tsx [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+    </MenuItem>
+  );
+};
+
+const WriterDropdown: React.FC<WriterDropdownProps> = ({
+  tableApps,
+}: WriterDropdownProps) => {
+  const tableAppName = tableApps[0].name.toLowerCase();
+
+  let image = '';
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/index.tsx:95]: Let's extract this into a helper function, something like `getImagePath(tableAppName)`.
I would also wrap it into a `useCallback` function(https://www.robinwieruch.de/react-usecallback-hook), as this gets executed on every component re-render.

File: frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/index.tsx [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+      break;
+    case DATABRICKS:
+      image = '/static/images/icons/logo-databricks.png';
+      break;
+    default:
+      image = '/static/images/icons/Application.svg';
+  }
+
+  tableApps.sort(
+    (a, b) => a.name.localeCompare(b.name) || a.id.localeCompare(b.id)
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/index.tsx:108]: I would extract this oneliner as `sortByNameOrId`

File: frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/index.tsx [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+  );
+  const appNames: string[] = [...new Set(tableApps.map((app) => app.name))];
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/index.tsx:110]: Nit:
```suggestion
  const appNames: string[] = [...new Set(tableApps.map(({name}) => name))];
```

File: frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/index.tsx [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+
+  const menuItems: JSX.Element[] = [];
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/index.tsx:112]: I usually don't use JSX.Element, I think React.ReactNode will work better.

File: frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/index.tsx [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+      menuItems.push(
+        <h5 className="writer-dropdown-menu-title">{sectionTitle}</h5>
+      );
+      menuItems.push(
+        ...tableApps
+          .filter(
+            (app) =>
+              app.name === name &&
+              ((app.kind && app.kind === kind) ||
+                (!app.kind && kind === PRODUCING))
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/index.tsx:131]: Could you summarize this predicate into something human readable?
Maybe `hasSameNameAndKind` or similar?

File: frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/index.tsx [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+      tableApps.filter((app) => app.name === name)
+    );
+    appKinds.forEach((kind, kindIdx) => {
+      const sectionTitle = isAirflowOrDatabricksApp(tableApps[0].name)
+        ? kind
+        : name + ' - ' + kind;
+      menuItems.push(
+        <h5 className="writer-dropdown-menu-title">{sectionTitle}</h5>
+      );
+      menuItems.push(
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/index.tsx:125]: Same comment about `push`, there isn't a good reason to use it in most of the cases.

File: frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/index.tsx [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+            (app) =>
+              app.name === name &&
+              ((app.kind && app.kind === kind) ||
+                (!app.kind && kind === PRODUCING))
+          )
+          .map((app) => {
+            const menuItemClassName =
+              ++appsDefinedCount === tableApps.length
+                ? 'writer-dropdown-menu-item last-item'
+                : '';
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/index.tsx:137]: Hmm... this was interesting, could you expand a bit why we do it?

File: frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/styles.scss [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+  .writer-dropdown-menu-title {
+    margin: $spacer-2 $spacer-2 0 $spacer-2;
+    line-height: $spacer-2;
+    text-transform: uppercase;
+    font-size: $menu-title-font-size;
+    white-space: nowrap;
+  }
+
+  .writer-dropdown-menu-item {
+    &.last-item {
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/styles.scss:36]: You might not need to add this class by using https://developer.mozilla.org/en-US/docs/Web/CSS/:last-child, give it a try!

File: frontend/amundsen_application/static/js/pages/TableDetailPage/index.tsx [hunk 16]
```diff
@@ -341,6 +343,50 @@ export class TableDetail extends React.Component<
...
+
+    if (airflowApps.length > 0) {
+      writerDropdowns.push(<WriterDropdown tableApps={airflowApps} />);
+    }
+    if (databricksApps.length > 0) {
+      writerDropdowns.push(<WriterDropdown tableApps={databricksApps} />);
+    }
+    if (remainingApps.length > 0) {
+      writerDropdowns.push(<WriterDropdown tableApps={remainingApps} />);
+    }
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/pages/TableDetailPage/index.tsx:385]: I think I see what you are trying to do here, but it feels non-idiomatic in the React world.
Check out this resource: https://reactjs.org/docs/conditional-rendering.html

We tend to favor the 'Inline If with Logical && Operator' method nowadays.

File: frontend/amundsen_application/static/js/pages/TableDetailPage/index.tsx [hunk 17]
```diff
@@ -341,6 +343,50 @@ export class TableDetail extends React.Component<
...
   }
 
+  renderTableWriterDropdowns(tableWriter, tableApps) {
+    const apps: TableWriter[] = [];
+    if (
+      (tableApps === null || tableApps.length === 0) &&
+      tableWriter === null
+    ) {
+      return null;
+    }
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/pages/TableDetailPage/index.tsx:353]:
```suggestion
const hasNoAppsOrWriter = (tableApps === null || tableApps.length === 0) &&
      tableWriter === null;
    if (hasNoAppsOrWriter) {
      return null;
    }
```

File: frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/styles.scss [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+
+    .menu-item-content {
+      line-height: $menu-item-content-height;
+      max-width: fit-content;
+      overflow: hidden;
+      text-overflow: ellipsis;
+      white-space: nowrap;
+    }
+  }
+}
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/styles.scss:57]: This looks great, awesome job!

File: frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/index.tsx [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+  const appNames: string[] = [...new Set(tableApps.map((app) => app.name))];
+
+  const menuItems: JSX.Element[] = [];
+  let appsDefinedCount = 0;
+  appNames.forEach((name, nameIdx) => {
+    const appKinds = getSortedAppKinds(
+      tableApps.filter((app) => app.name === name)
+    );
+    appKinds.forEach((kind, kindIdx) => {
+      const sectionTitle = isAirflowOrDatabricksApp(tableApps[0].name)
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/index.tsx:119]: I wonder if it would be better to have just different menuItem components for the different apps.
Even if there is some duplication on the JSX markup, we would probably avoid some of the logic I we are using on this PR.

Think a branching like:
```
<Dropdown.Menu className="writer-dropdown-menu">
        {
  if (isAirflow) {
    return <AirFlowMenuItem />;
  } else if (isDatabricks) {
    return <DataBricksMenuItem  />;
  } else {
    return <GenericMenuItem />;
  }}
 </Dropdown.Menu>
```

File: frontend/amundsen_application/static/images/icons/Application.svg (hunk 5 above)
Comment (kristenarmes) [reply frontend/amundsen_application/static/images/icons/Application.svg]: This icon looks like a gear with a square around it to represent generic applications (designed by Knowl). I renamed the file to be lowercase now, I noticed there was a mixture of some upper and some lower for the icon naming and chose wrong 😄

File: frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/index.tsx (hunk 14 above)
Comment (kristenarmes) [reply frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/index.tsx]: I realized after you pointed this out that this was unnecessary - I was originally adding a `margin-bottom` to the last item in the list, but instead I just removed this extra work and added padding to the bottom of the dropdown itself

File: frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/index.tsx (hunk 19 above)
Comment (kristenarmes) [reply frontend/amundsen_application/static/js/pages/TableDetailPage/WriterDropdown/index.tsx]: This is a good suggestion, it will make it easier/cleaner too if anyone else in the future wants to add special styling for a separate application. I'll make this change in the remaining work on this PR

**Comment by dkunitsk** [comment]
Some thoughts after trying this out internally:
- could the dropdown be shifted left (i.e. the right edge of the dropdown aligned with the chevron) and be made wider so that more of the text is visible
- I'm thinking 'not available' is actually a better and more clear placeholder than 'none' for task_id.

File: frontend/amundsen_application/static/js/pages/TableDetailPage/ApplicationDropdown/index.tsx [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+import './styles.scss';
+
+export interface ApplicationDropdownProps {
+  tableApps: TableApp[];
+}
+
+const getImagePath = (tableAppName) => {
+  switch (tableAppName) {
+    case AIRFLOW.toLowerCase():
+      return '/static/images/airflow.jpeg';
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/pages/TableDetailPage/ApplicationDropdown/index.tsx:29]: Nit: I would move this strings into constants, something like:
const AIRFLOW_LOGO_PATH
const DATABRICKS_LOGO_PATH
...

File: frontend/amundsen_application/static/js/pages/TableDetailPage/ApplicationDropdown/index.tsx [hunk 21]
```diff
@@ -0,0 +1,139 @@
...
+  return [...producingKind, ...consumingKind, ...remainingKinds];
+};
+
+const handleClick = (event) => {
+  event.stopPropagation();
+  logClick(event);
+};
+
+const getDropdownMenuContents = (tableApps) => {
+  if (tableApps[0].name.toLowerCase() === AIRFLOW.toLowerCase()) {
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/pages/TableDetailPage/ApplicationDropdown/index.tsx:77]: Nit:
const isAirFlow = tableApps[0].name.toLowerCase() === AIRFLOW.toLowerCase();
const isDataBricks = tableApps[0].name.toLowerCase() === DATABRICKS.toLowerCase();

if (isAirFlow) { ...

File: frontend/amundsen_application/static/js/pages/TableDetailPage/ApplicationDropdown/index.tsx [hunk 22]
```diff
@@ -0,0 +1,139 @@
...
+      hasSameNameAndKind={hasSameNameAndKind}
+      handleClick={handleClick}
+    />
+  );
+};
+
+const ApplicationDropdown: React.FC<ApplicationDropdownProps> = ({
+  tableApps,
+}: ApplicationDropdownProps) => {
+  if (tableApps === null || tableApps.length === 0) return null;
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/pages/TableDetailPage/ApplicationDropdown/index.tsx:110]: We don't do one liners like these, is an historic pattern we follow:
```suggestion
  if (tableApps === null || tableApps.length === 0) {
      return null;
  }
```

File: frontend/amundsen_application/static/js/pages/TableDetailPage/index.tsx [hunk 23]
```diff
@@ -341,6 +343,50 @@ export class TableDetail extends React.Component<
...
+    if (hasNoAppsOrWriter) {
+      return null;
+    }
+    if (tableApps !== null && tableApps.length > 0) {
+      apps = [...tableApps];
+    }
+    if (
+      tableWriter !== null &&
+      !apps.some((app) => app.id === tableWriter.id)
+    ) {
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/pages/TableDetailPage/index.tsx:359]: Would love to have another 'isNotWriterAnd...' here too

**Review by Golodhros — COMMENTED** [review]
Looks much better!

Only some minor comments

**Review by Golodhros — APPROVED** [review]
LGTM

Thanks!
